core/src/main/java/org/mule/runtime/core/api/policy/PolicyChain.java

```diff
@@ -79,20 +134,137 @@ public void stop() throws MuleException {
+    private List<Processor> afterNext;
+
+    private Scheduler beforeScheduler;
+    private Scheduler afterScheduler;
+    private Scheduler cpuLiteBeforeScheduler;
+    private Scheduler cpuLiteAfterScheduler;
+
+    public SourcePolicyProcessingStrategy(SchedulerService schedulerService, SchedulerConfig schedulerBaseConfig,
+                                          String schedulersNamePrefix, List<Processor> beforeNext, List<Processor> afterNext) {
+      this.schedulerService = schedulerService;
+      this.schedulerBaseConfig = schedulerBaseConfig;
+      this.schedulersNamePrefix = schedulersNamePrefix;
+      this.beforeNext = beforeNext;
+      this.afterNext = afterNext;
+    }
+
+    @Override
+    public void start() throws MuleException {
+      beforeScheduler = schedulerBasedOnProcessingTypes(beforeNext, schedulersNamePrefix + "-before");
+      afterScheduler = schedulerBasedOnProcessingTypes(afterNext, schedulersNamePrefix + "-after");
+      cpuLiteBeforeScheduler =
+          schedulerService.cpuLightScheduler(schedulerBaseConfig.withName(schedulersNamePrefix + "-cpuLite-before"));
+      cpuLiteAfterScheduler =
+          schedulerService.cpuLightScheduler(schedulerBaseConfig.withName(schedulersNamePrefix + "-cpuLite-after"));
+    }
+
+    private Scheduler schedulerBasedOnProcessingTypes(List<Processor> processors, String schedulerName) {
+      Set<ProcessingType> types = processors.stream().map(p -> p.getProcessingType()).collect(toSet());
+
+      SchedulerConfig config = schedulerBaseConfig.withName(schedulerName);
+      if (types.stream().anyMatch(p -> BLOCKING.equals(p) || IO_RW.equals(p))) {
+        return schedulerService.ioScheduler(config);
+      } else if (types.stream().anyMatch(p -> CPU_INTENSIVE.equals(p))) {
+        return schedulerService.cpuIntensiveScheduler(config);
+      } else {
+        return null;
+      }
+    }
+
+    @Override
+    public void stop() throws MuleException {
+      if (beforeScheduler != null) {
+        beforeScheduler.stop();
+      }
+      if (afterScheduler != null) {
+        afterScheduler.stop();
+      }
+      if (cpuLiteBeforeScheduler != null) {
+        cpuLiteBeforeScheduler.stop();
+      }
+      if (cpuLiteAfterScheduler != null) {
+        cpuLiteAfterScheduler.stop();
+      }
+    }
+
+    @Override
+    public Sink createSink(FlowConstruct flowConstruct, ReactiveProcessor pipeline) {
+      return new StreamPerEventSink(pipeline, createOnEventConsumer());
+    }
+
+    @Override
+    public ReactiveProcessor onPipeline(ReactiveProcessor pipeline) {
+      return publisher -> Flux.from(publisher)
+          .publishOn(beforeScheduler != null ? fromExecutor(beforeScheduler) : immediate())
+          .transform(pipeline);
+    }
+
+
+    @Override
+    public ReactiveProcessor onProcessor(ReactiveProcessor processor) {
+      ReactiveProcessor reactiveProcessor = handleCpuLiteAsync(processor);
+
+      if (!afterNext.isEmpty()
```

Comment:
It would be a lot easier to understand this if this logic was extracted into a private method, maybe as a predicate.